src/Conversion/ONNXToKrnl/Tensor/NonZero.cpp

```diff
@@ -0,0 +1,138 @@
+    createKrnl.iterateIE(countLoopDef, countLoopDef, lbs, ubs,
+        [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
+          MathBuilder createMath(createKrnl);
+          Value val = createKrnl.load(X, loopInd);
+          Value eqCond = createMath.eq(val, zero);
+          Value zeroOrOne = createMath.select(eqCond, iZero, iOne);
+          Value counter = createKrnl.load(nonzeroCount, {});
+          Value newCount = createMath.add(counter, zeroOrOne);
+          createKrnl.store(newCount, nonzeroCount);
+        });
+
+    // Insert an allocation and deallocation for the result of this operation.
+    Value numberOfZeros = createKrnl.load(nonzeroCount, {});
+    SmallVector<IndexExpr, 2> dimExprs;
+    dimExprs.emplace_back(litRank);
+    dimExprs.emplace_back(DimIndexExpr(numberOfZeros));
+    bool insertDealloc = checkInsertDealloc(op);
+    Value resMemRef = insertAllocAndDeallocSimple(
+        rewriter, op, resMemRefType, loc, dimExprs, insertDealloc);
+
+    // When computing indices of nonzero value, keep trace of current indices
+    // for each dimension so that we know exactly where to store in the result.
+    Value pos =
+        insertAllocAndDealloc(MemRefType::get({xRank}, rewriter.getIndexType()),
+            loc, rewriter, /*insertDealloc=*/true);
+    ValueRange posLoopDef = createKrnl.defineLoops(1);
+    createKrnl.iterateIE(posLoopDef, posLoopDef, {litZero}, {litRank},
+        [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
+          createKrnl.store(iZero, pos, loopInd);
+        });
+
+    // Iterate over the input in row-major order to get indices of nonzero
+    // values.
+    ValueRange mainLoopDef = createKrnl.defineLoops(xMemRefType.getRank());
+    createKrnl.iterateIE(mainLoopDef, mainLoopDef, lbs, ubs,
+        [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
+          MathBuilder createMath(createKrnl);
+          Value val = createKrnl.load(X, loopInd);
+          Value eqCond = createMath.eq(val, zero);
+          Value zeroOrOne = createMath.select(eqCond, iZero, iOne);
+          for (int64_t i = 0; i < xRank; ++i) {
+            SmallVector<IndexExpr, 1> posInd;
+            posInd.emplace_back(LiteralIndexExpr(i));
+            // Load the current position along dimension i.
+            Value currentPos = createKrnl.loadIE(pos, posInd);
+            // Load the output at the current position along dimension i.
+            SmallVector<IndexExpr, 2> resInd;
+            resInd.emplace_back(LiteralIndexExpr(i));
+            resInd.emplace_back(DimIndexExpr(currentPos));
+            // If value is nonzero, update the output with the value's index.
+            // Otherwise, keep the output unchanged.
+            Value oldIndex = createKrnl.loadIE(resMemRef, resInd);
+            Value newIndex =
+                rewriter.create<IndexCastOp>(loc, loopInd[i], resElementType);
+            newIndex = createMath.select(eqCond, oldIndex, newIndex);
```

Comment:
Pretty smart, but could we possibly have an unconditional write after the last reserved position?
It might be a cause to use an affine if-then... I have a copy that moved it from affine (non-exportable) to krnl... listed here if useful to you.

KrnlHelper.hpp and cpp

``` cpp
  void ifThenElse(IndexExprScope &scope, SmallVectorImpl<IndexExpr> &conditions,
      function_ref<void(KrnlBuilder &)> thenFn,
      function_ref<void(KrnlBuilder &)> elseFn);
      
  private:
    void appendToBlock(Block *block, function_ref<void(ValueRange)> builderFn);
```

``` cpp
void KrnlBuilder::ifThenElse(IndexExprScope &scope,
    SmallVectorImpl<IndexExpr> &conditions,
    function_ref<void(KrnlBuilder &)> thenFn,
    function_ref<void(KrnlBuilder &)> elseFn) {
  int64_t rank = conditions.size();
  SmallVector<AffineExpr, 4> affineCond;
  bool allTrue = true;
  bool allFalse = true;
  for (IndexExpr c : conditions) {
    assert(c.isAffine() && "conditions expected to be affine");
    affineCond.emplace_back(c.getAffineExpr());
    if (c.isLiteral()) {
      // Condition can be evaluated at compile time, as it is a literal.
      if (c.getLiteral() < 0) // Inequality is expr >= 0, test if false.
        allTrue = false;
      if (c.getLiteral() >= 0) // Inequality is expr >= 0, test if true.
        allFalse = false;
    } else {
      // Condition cannot be evaluated at compile time
      allTrue = allFalse = false;
    }
  }
  SmallVector<bool, 4> isEq(rank, false);
  auto inset = IntegerSet::get(
      scope.getNumDims(), scope.getNumSymbols(), affineCond, isEq);
  SmallVector<Value, 8> dimAndSymbolList;
  scope.getDimAndSymbolList(dimAndSymbolList);
  auto ifOp = b.create<AffineIfOp>(loc, inset, dimAndSymbolList, true);
  Block *thenBlock = ifOp.getThenBlock();
  Block *elseBlock = ifOp.getElseBlock();
  if (!allFalse) {
    appendToBlock(thenBlock, [&](ValueRange args) {
      KrnlBuilder createKrnl(b, loc);
      thenFn(createKrnl);
    });
  }
  if (!allTrue) {
    appendToBlock(elseBlock, [&](ValueRange args) {
      KrnlBuilder createKrnl(b, loc);
      elseFn(createKrnl);
    });
  }
}

void KrnlBuilder::appendToBlock(
    Block *block, function_ref<void(ValueRange)> builderFn) {
  OpBuilder::InsertionGuard guard(b);
  if (block->empty() ||
      !block->back().mightHaveTrait<OpTrait::IsTerminator>()) {
    b.setInsertionPointToEnd(block);
  } else
    b.setInsertionPoint(&block->back());
  builderFn(block->getArguments());
}
```